typedb/concept/answer/concept_map.py

```diff
@@ -19,138 +19,171 @@
 # under the License.
 #
 
-from typing import Mapping, Dict, Tuple
+from __future__ import annotations
 
-import typedb_protocol.common.answer_pb2 as answer_proto
+from typing import Mapping, Iterator, TYPE_CHECKING
+
+from typedb.native_client_wrapper import concept_map_get_variables, string_iterator_next, concept_map_get_values, \
+    concept_iterator_next, concept_map_get, concept_map_get_explainables, concept_map_to_string, concept_map_equals, \
+    explainables_get_relation, explainables_get_attribute, explainables_get_ownership, \
+    explainables_get_relations_keys, explainables_get_attributes_keys, explainables_get_ownerships_keys, \
+    string_pair_iterator_next, explainables_to_string, explainables_equals, explainable_get_conjunction, \
+    explainable_get_id
 
 from typedb.api.answer.concept_map import ConceptMap
-from typedb.api.concept.concept import Concept
-from typedb.common.exception import TypeDBClientException, VARIABLE_DOES_NOT_EXIST, NONEXISTENT_EXPLAINABLE_CONCEPT, \
-    NONEXISTENT_EXPLAINABLE_OWNERSHIP
-from typedb.concept.proto import concept_proto_reader
+from typedb.common.exception import TypeDBClientExceptionExt, ILLEGAL_STATE, MISSING_VARIABLE, \
+    NONEXISTENT_EXPLAINABLE_CONCEPT, NONEXISTENT_EXPLAINABLE_OWNERSHIP, NULL_NATIVE_OBJECT, VARIABLE_DOES_NOT_EXIST
+from typedb.common.iterator_wrapper import IteratorWrapper
+from typedb.common.native_object_mixin import NativeObjectMixin
+from typedb.concept import concept_factory
+
+if TYPE_CHECKING:
+    from typedb.api.concept.concept import Concept
+    from typedb.native_client_wrapper import ConceptMap as NativeConceptMap, Explainables as NativeExplainables, \
+        Explainable as NativeExplainable
+
+
+def _not_blank_var(var: str) -> str:
+    if not var or var.isspace():
+        raise TypeDBClientExceptionExt.of(MISSING_VARIABLE)
+    return var
 
 
-class _ConceptMap(ConceptMap):
+class _ConceptMap(ConceptMap, NativeObjectMixin):
 
-    def __init__(self, mapping: Mapping[str, Concept], explainables: ConceptMap.Explainables = None):
-        self._map = mapping
-        self._explainables = explainables
+    def __init__(self, concept_map: NativeConceptMap):
+        if not concept_map:
+            raise TypeDBClientExceptionExt(NULL_NATIVE_OBJECT)
+        self.__native_object = concept_map
```

Comment:
If we pass the native handle up to `NativeObjectMixin.__init__(concept_map)`, then we don't need the property in every subtype.

If we want to preserve the return type to be specific, like `NativeConceptMap`, then we use a generic on `NativeObjectMixin[T]`.